#728 accessibility theming

atarix83 opened the pull request:
## References
* Fixes #728 
* Fixes #1030 
* Fixes #976 
* Fixes #1020 

## Description
This PR is created in order to define [UI design guidelines and accessibility principles](https://wiki.lyrasis.org/display/DSPACE/DSpace+UI+Design+principles+and+guidelines) to be followed by DSpace's contributors. It also provides a rough clean-up of the existing components/styles to solve some accessibility/usability issues. Some follow-up usability/accessibility issues that can be covered by this PR will be created soon.

## Instructions for Reviewers
To review this PR please navigate as many dspace's functioanlity as possible to verify that the UI is more consistent and follows the above mentioned guidelines.
Check also if bugs reported in the issue #1020 and #1030 are resolved

List of changes in this PR:
* Splitted the header, in two parts. The user account related functions are isolated in a second smaller menu because, in our experience, many DSpace institutions have too much items inside the main menu. This will lead inevitably to a navbar that is too crowded to properly accommodate other objects.
* Created a double footer
* Minor cleanup to CSS or HTML to ensure the above mentioned guidelines
* to solve #976 the page title is now created by concatenating automatically the `DSpace Angular ::`  prefix
* set https://brailleinstitute.org/freefont as default font 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

tdonohue reviewed (changes requested):
@atarix83 : I gave this a really quick look today.  Overall, the changes seem good (I haven't done a full code review yet).  However, I was surprised to see a number of unexpected styling changes (CSS) that seem to have changed the default look-and-feel more significantly.  

Some of those unexpected changes include:

1. A bright blue top header and lower footer, both of which seem to move away from our default Bootstrap look & feel. 
2. A DSpace green line below the header section & a DSpace green colored link in the header. These should both be in our eventual theme, as they are definitely not default Bootstrap.
3. A gray gradient now appears on the homepage, and this is also new.
4. The blue color of all button & links has been changed.

To clarify, the scope of ticket #728 was limited to HTML specific changes to header/footer , accessibility and alignment.  No new theme work should occur in this PR, which means there should not be any CSS changes **unless** they are required to fix accessibility issues (for example color contrast, etc).  But, several of the changes seem unrelated to accessibility.

For others, here's a screenshot of the homepage when running this PR:
![homepage](https://user-images.githubusercontent.com/483997/111546509-cacf2b80-8745-11eb-91e2-baaa3cbe4a6b.png)

atarix83 commented:
@tdonohue 
  

- The 'user profile bar' and, generally, the color associated with primary actions (such as links and buttons) have been rendered in a darker blue to ensure a greater contrast (and consequently a clearer separation) with the 'main navigation bar' and background elements. This ensures high standards of accessibility according to the rule '1.4 Distinguishable: Make it easier for users to see and hear content including separating foreground from background' of WCAG 2.0.
    

- For the same reason, a border has been applied below the 'main navigation bar'. This element creates a clear distinction between the header and the main page, avoiding any confusion with other possible menus below. see also (https://github.com/DSpace/dspace-angular/issues/1070)

artlowel commented:
>The 'user profile bar' and, generally, the color associated with primary actions (such as links and buttons) have been rendered in a darker blue to ensure a greater contrast (and consequently a clearer separation) with the 'main navigation bar' and background elements. This ensures high standards of accessibility according to the rule '1.4 Distinguishable: Make it easier for users to see and hear content including separating foreground from background' of WCAG 2.0.

I'm not convinced the user profile bar needs to be _just so_ prominent. In this PR it really jumps out at you as if it was the most important thing on the page. While the features in that bar are not needed for most users; they only need the public parts of DSpace, especially now you've moved the search bar down (I like that change). 

I think simply switching the link color to the darker blue and leaving the background white would have made that menu prominent enough. If you insist on using an inverted color-scheme for this bar, I'd go for a less saturated color. Bootstrap's default inverted navbar colors seem the obvious choice.

tdonohue commented:
@atarix83 : Again, we should **not** be using the DSpace green color in this PR, as that color is being reserved for our sample theme.  As discussed previously, the "base theme" (which this PR creates) should keep to a simply color scheme (which is Bootstrap like)

I ran a few contrast tests on my end (in chrome), and it looks like the primary contrast issue is that currently (e.g. on https://demo7.dspace.org) the default link color is $cyan (#2790B0) instead of our current (darker) $blue (#2B4E72).  So, if you switch the link color to be $blue, that'll fix the contrast issues (without having to modify the current colors in `_bootstrap_variables.scss` at all).  The same should work fine for the login header. If that $blue doesn't look good in the header, a gray header is fine (there's no need for it to be blue, it could just be Bootstrap's inverted header as @artlowel noted).

As for the DSpace green line, it should be removed in favor of a simple gray or dark blue.  The gradient should also be removed, as a gradient makes color contrast even harder to achieve.  If you change the link color, you likely can keep the "news" section gray (like it is on https://demo7.dspace.org), or switch it to a lighter gray or white background if needed.

Overall, as noted, I want to ensure this PR is not significantly changing the color scheme from https://demo7.dspace.org, and that color scheme is primarily gray / dark blue / dark gray (almost black) / white.   Also, I believe this color scheme will make it easier to provide good color contrast.  

(As a sidenote, it's worth being aware that we unfortunately _cannot_ revert to default Bootstrap colors, because Bootstrap itself notes that it's default colors [do not provide proper color contrast on light backgrounds](https://getbootstrap.com/docs/4.0/getting-started/accessibility/#color-contrast).)

If you have questions on the requested changes, or need further guidance, let me know.  Thanks!

atarix83 commented:
> I'm not convinced the user profile bar needs to be quite so prominent. In this PR it really jumps out at you as if it was the most important thing on the page. While the features in that bar are not needed for most users; they only need the public parts of DSpace, especially now you've moved the search bar down (I like that change).
I think simply switching the link color to the darker blue and leaving the background white would have made that menu prominent enough. If you insist on using an inverted color-scheme for this bar, I'd go for a less saturated color. Bootstrap's default inverted navbar colors seem the obvious choice.

@artlowel thanks for your feedback, I'll try to change the background color to white

> Again, we should not be using the DSpace green color in this PR, as that color is being reserved for our sample theme. As discussed previously, the "base theme" (which this PR creates) should keep to a simply color scheme (which is Bootstrap like)

In our opinion the default theme should remain as simple as possible and easy to modify in custom themes, but it should also be attractive enough to be used as it is, because we know that many people don't change it. Furthermore when people compare DSpace with digital commons, or other similar products, they look at the demo that uses the default theme.

We also think that [the sample theme](https://github.com/DSpace/dspace-angular/pull/1063) should be just an example on how to produce an institutional custom theme and should not be more appealing than the default one. Otherwise we run the risk that people start to use it instead of the default one and that would require more effort to maintain both in the future.

So if it's possible to make small adjustments (for example adding a colored border) in this PR that help to make the interface more attractive , we think it could be very useful.

> I ran a few contrast tests on my end (in chrome), and it looks like the primary contrast issue is that currently (e.g. on https://demo7.dspace.org) the default link color is $cyan (#2790B0) instead of our current (darker) $blue (#2B4E72). So, if you switch the link color to be $blue, that'll fix the contrast issues (without having to modify the current colors in _bootstrap_variables.scss at all). The same should work fine for the login header. If that $blue doesn't look good in the header, a gray header is fine (there's no need for it to be blue, it could just be Bootstrap's inverted header as @artlowel noted).
As for the DSpace green line, it should be removed in favor of a simple gray or dark blue. The gradient should also be removed, as a gradient makes color contrast even harder to achieve. If you change the link color, you likely can keep the "news" section gray (like it is on https://demo7.dspace.org), or switch it to a lighter gray or white background if needed.
Overall, as noted, I want to ensure this PR is not significantly changing the color scheme from https://demo7.dspace.org, and that color scheme is primarily gray / dark blue / dark gray (almost black) / white. Also, I believe this color scheme will make it easier to provide good color contrast.
(As a sidenote, it's worth being aware that we unfortunately cannot revert to default Bootstrap colors, because Bootstrap itself notes that its default colors do not provide proper color contrast on light backgrounds.)

we should avoid to use the same color for primary actions, based on `$blue`, and for links and info actions that are bases on the `$cyan`. so if the color used for `$cyan` is not good enough, we should change it but we shouldn't use the same used for `$blue`

artlowel commented:
>In our opinion the default theme should remain as simple as possible and easy to modify in custom themes, but it should also be attractive enough to be used as it is, because we know that many people don't change it. Furthermore when people compare DSpace with digital commons, or other similar products, they look at the demo that uses the default theme.
>
>We also think that the sample theme should be just an example on how to produce an institutional custom theme and should not be more appealing than the default one. Otherwise we run the risk that people start to use it instead of the default one and that would require more effort to maintain both in the future.

We went over this multiple times in meetings…. 

We agreed that the scope for #728 was to do accessibility improvements, not theming. The goal for the base theme is to be plain bootstrap. No-one is likely to use that as is, in almost all cases there will be another theme on top of that. 

The dspace default theme is the subject of #727. It will be a small layer of css on top, and not much more. When institutions build their own theme, they can choose to start from the dspace theme if they like. But if all that style is kept out of the base components they have a blank canvas to start from, in case they want to diverge from the dspace theme more radically. That's why we agreed not to put any unnecessary stying in the base components. Not just recently, but that's a rule we've had since the inception of dspace 7.

tdonohue commented:
@atarix83: Just for clarity...the DSpace green color **must** be removed from this PR.  We've discussed this before, and I'm not willing to compromise on this. Please remove that color from the base color scheme. To clarify, my rule for this PR is

 **The base color scheme should be mostly Bootstrap defaults, with any necessary tweaks to make specific colors more accessible on a white/light background.**

This means that if we need to change the value of $cyan or $blue, I'm OK with that **provided** that the color change is only for accessibility reasons.  It is not appropriate in this PR to change a color just to change the color scheme...and this PR currently changes $blue for no apparent reason (as I noted, the current $blue is already passing accessibility tests, I believe).

My comments with regards to the links were not to say that $cyan and $blue should be *identical colors*.  I was noting instead that we could potentially change the links to use the $blue value _instead of_ $cyan, so that $blue is the main color used instead of $cyan.  However, if you'd prefer to change $cyan to be more accessible, that's OK.  However, in this PR you appear to have changed $cyan to #4272d...and at least on my screen #4272d looks **blue** and not **cyan**.  If you wanted an accessible $cyan, something like #006666 may be good enough (other cyan like colors are fine, but we shouldn't change `$cyan` to look blue.. as that's the purpose of `$blue`)

atarix83 commented:
@artlowel @tdonohue  I reverted the original color schema and changing only the `$cyan` color with the one suggested by Tim. The top header is now simply white and I've also removed the gradient in the home page. Please let me know if there are other changes to apply

tdonohue commented:
Apologies, @atarix83 .  I've accidentally created some merge conflicts here with recent mergers.  When you get a chance, could you update/rebase this PR on latest `main`?

artlowel reviewed (changes requested):
Thanks @atarix83!

I have some feedback below. While there are a few irrefutable problems, a lot of it is subjective as well, which is inescapable in reviews of PRs involving style.

#### Header
Thanks for changing the background color for the top header bar. I'd also remove the border separating the two parts. 

It's strange that the section containing the logo isn't in the `<header>` tag anymore. That looks like a step backwards accessibility-wise, as it will make it more difficult for screen readers to make sense of the site. 

While I like the more compact look of the header for the specific case of the dspace logo, I know from experience that most institutions will have a rectangular logo that'll be a lot wider. If they simply replace the square dspace one with it, it will push the navigation to the right

There is a point at around 750px wide (you can set that in the device toolbar in the dev tools) where the header breaks:
![Screen Shot 2021-03-30 at 11 38 35](https://user-images.githubusercontent.com/1567693/112968339-7d42bd80-914c-11eb-9fa8-fde52d5d86fe.png)


#### Footer
I think it's good to have a template for the top footer, however a footer with a bunch of lorem ipsum content in is guaranteed to have to be modified by everybody. It would ensure you couldn't use dspace out of the box. So I would comment it out by default. I'd also split the two footers into separate angular components.

I think the bottom footer could use more padding/spacing, and is also too prominent at the moment. I'd use a less saturated background color e.g. $dark or $gray-800, and change the foreground color to $gray-600.

#### Font
I don't like the font. I know @tdonohue suggested it, but I think it goes too far. The zeroes really put me off (for example the copyright dates in the footer, or the last modified date on the edit item status page), and the use of the serif capital i in a sans serif font looks very strange (e.g. 'Item Type' in the facets on the search page). The font also seems to be the reason why text is no longer centered vertically on buttons and text fields etc.

I'm in favor of simply using the default bootstrap font-stack. Which uses the default fonts for the OS, that have been designed by Microsoft, Google and Apple with readability in mind already, and falls back to Helvetica and Arial, both also accessible fonts in their own right.

#### Homepage news link color
Homepage news uses a different link color from everything else. I'd change it to use the default link color.

#### Buttons
I don't like the change where the edit button on dso pages is now an outline. It was given a dark color on purpose to remind people of the sidebar and communicate that it is a feature for authenticated users only.

The button to persist changes on edit pages has been called "Submit" in most places, "Save" in others (e.g. edit item metadata). I think "Save" is clearest and would use that everywhere. On the profile page it is called "Update Profile", I'd call it "Save" there as well.

The cancel button has been renamed "← Back" in some places (e.g. edit eperson) "←  Return" in others (e.g. edit item/collection/community". I think "← Back" is the clearest and would use that everywhere.

The save button is the rightmost button almost everywhere. Except in some cases where there is a delete or discard option (e.g. edit item metadata, edit eperson). I'd expect the primary action button to be the rightmost one, the back button to be the leftmost one, and any other buttons to be in between

A number of pages haven't gotten their buttons revisited, and still use diverging terms, or different positioning. Here's a number I came across:
- all the pages linking from the edit item - status tab (withdraw, permanently delete, etc)
- Item mapper on the edit item page has both a cancel and a return button
- Import metadata
- Curation tasks
- Processes

tdonohue reviewed (changes requested):
@atarix83 & @artlowel :  In order to move this PR along quickly, I'd  recommend we make ONLY the following changes to this PR. Any other changes can be delayed for a future fixes -- **remember** this is not the final theme or accessibility work, there will be additional accessibility work after the accessibility analysis performed during Testathon.   So, I'd encourage us to consider this PR a **starting point** and we can continue to improve the accessibility / layout / consistency in future PRs

**These are tasks I feel should be done in this PR:**
1. **Remove the font.**  After further analysis, I agree with @artlowel .  Let's just keep this simple for now and use native, system fonts. It looks like that means we can simply remove the changes to `/src/assets/fonts` and remove the definition of `$font-family-sans-serif` (in `_bootstrap_variables.scss`), since Bootstrap uses native fonts by default: https://getbootstrap.com/docs/4.0/content/reboot/#native-font-stack 
2. Remove the lines between the top part of the header and the bottom.  Just change these variables in `_custom_variables.scss`:
    ```
    --ds-header-navbar-border-top-color: #{$white};
    --ds-header-navbar-border-bottom-color: #{$white};
    ```
3. Change homepage link color. This is a simple variable change too
    ```
    --ds-home-news-link-color: #{$cyan};
    --ds-home-news-link-hover-color: #{darken($cyan, 15%)};
    ```
4. It appears we need to darken the other links in that header nav-bar.  I'm using this [Chrome extension](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US), and it tells me the "Communities & Collection" / "All of DSpace" and "Statistics" links all have insufficient contrast.  We may want to create a `ds-navbar-link-color` variable and set it to $cyan (or a dark gray / black) for now.
5. (Optional...could be cleaned up later) That same tool shows that the descriptions under any Community links in the "Communities in DSpace" section also have insufficient contrast, as does the text "Search the repository" in the main search box.  I'm not sure if these are very easy to solve though, so we could follow-up this PR with more small accessibility changes.

**These are tasks that can be done later:**
* Ensure the entire header is in the `<header>` element.  I agree with @artlowel on this, but this looks like a bigger task as it may require removing the HeaderNavBarHeaderComponent (which currently wraps the ds-header and ds-navbar....this isn't needed if the ds-navbar is within the ds-header)
* Button changes requested by @artlowel .  I feel this is a bit of a larger task too, and it might be easily done in a followup PR.  I do agree with his suggestions, but I don't want this PR to wait much longer in order to implement them.
* Suggested footer changes.  Again, good ideas, just not high priority for this PR & I'm worried they'd delay this PR too long.

Overall, I'd like to minimize the required changes in this PR, so that we can free up work on #1063 (which needs to be freed up if it's going to be ready in time for Beta 5)

artlowel commented:
@tdonohue I can agree with that apart from two things:

- I would comment out the lorem ipsum footer now. I created https://github.com/DSpace/dspace-angular/issues/1080 to address that same issue for the privacy statement and end-user agreement.
- The color of the edit button on DSO pages is a change made by this PR. I'd just roll it back, rather than change it now, and create a ticket to change it back later.

atarix83 commented:
@artlowel the reason we changed the edit button on dso page is that we tried to give common style rules when using buttons and we tried to uniform all over the application. The rules defined are liste in the section `UI Elements` of [the documentation](https://wiki.lyrasis.org/display/DSPACE/DSpace+UI+Design+principles+and+guidelines). 
In this specific case that button follow the rule `For the ‘anchor’ ( the ‘<a>’ element) that uses the ‘btn’ Bootstrap CSS class always use btn-outline-primary.` 
Before to revert the change, we can of course discuss on the rules defined in order to define a new rule eventually, but this should be consistent within the application.

artlowel commented:
In that case I would add a rule that UI elements on public pages that are only visible to users with elevated privileges use an inverted color scheme. That was the idea behind the dark sidebar, and this button.

atarix83 commented:
ok this makes sense. 

regarding the footer I'm not really sure about your suggestion, should I remove the footer content and leave only the container? something like this ? 

![Schermata da 2021-03-31 10-48-56](https://user-images.githubusercontent.com/2486489/113117552-da9f4300-920e-11eb-8f4c-e66759493f01.png)

artlowel commented:
No, I'd comment out the container as well, like this:

![image](https://user-images.githubusercontent.com/1567693/113118166-7df05800-920f-11eb-81db-b96fe806ec04.png)

That way people don't need to do anything if they don't need it, but people that want it can simply uncomment it and add their own links.

atarix83 commented on src/app/navbar/navbar.component.scss:
```diff
@@ -29,7 +31,18 @@ nav.navbar {
+}
+
+a.navbar-brand img {
+    max-height: 40px;
```
fixed

atarix83 commented:
@tdonohue  @artlowel 
I've addressed all tasks should be done in this PR suggested by Tim. I've also: 
- change the dso edit button style and updated the documentation with the rule suggested by Art
- disabled the top footer 
- try to ensure the save and back labels are used uniformly everywhere

Regarding the issue 
> (Optional...could be cleaned up later) That same tool shows that the descriptions under any Community links in the "Communities in DSpace" section also have insufficient contrast, as does the text "Search the repository" in the main search box. I'm not sure if these are very easy to solve though, so we could follow-up this PR with more small accessibility changes.

It is relatively simple to change because it consists to edit the following bootstrap variable :
- $text-muted 
- $input-placeholder-color
- $input-color

the issue is to decide which color to use. I've tried with the following values, but the result is that it's hard to discern from input placeholder and input value. If I use a ligheter color for the placeholder ($gray-600) the result it's the same as the current situation
```
$text-muted : $gray-700;
$input-placeholder-color: $gray-700;
$input-color: $gray-900;
```

Finally, only a comment on how header result without the border 
![Schermata da 2021-03-31 12-43-13](https://user-images.githubusercontent.com/2486489/113136782-0d076b00-9224-11eb-8f65-bac75799e3de.png)

personally I don't like it, because it seems slightly confused, indeed the aim of the border was to have a logical separation between the two different bars. If we don't want this logical separation I'd put all only in one navbar, probably something like this

![Schermata da 2021-03-31 13-30-25](https://user-images.githubusercontent.com/2486489/113137907-6cb24600-9225-11eb-8b34-d2cf316a0663.png)

tdonohue commented:
@atarix83 : Thanks for the updates & it is looking good now (based on the screenshots).  I'll try and install this later today to give you feedback on the color of the search box text.

Regarding your header screenshots, in the scope of this work, I had originally suggested a **single navbar** in the header (if you look back to the original ticket, look at "Scope of Ticket", number 4, second bullet) as your second screenshot shows.   So, if you also prefer a single navbar, then I'm OK with you switching to that in this PR (at least for now).  You could even look at putting everything in the `<header>` element, which would resolve another of @artlowel 's points of feedback.

artlowel commented on src/app/root/root.component.html:
```diff
@@ -10,14 +10,12 @@
+        <ds-loading [showMessage]="false"></ds-loading>
       </div>
-      <router-outlet></router-outlet>
+      <router-outlet *ngIf="!isLoading"></router-outlet>
```
@atarix83 While I think this is a good change, I think it's a better idea to _hide_ the router-outlet while loading. `*ngIf` doesn't render it at all, which means that only after isLoading turns false all the components that depend on the route can run _their_ onInit methods etc.

If we hide it instead, that can happen in the background. For the user it looks the same.
```suggestion
      <router-outlet [class.d-none]="isLoading"></router-outlet>
```

atarix83 commented:
@artlowel @tdonohue 
I committed changes in order to have navbar in one single row, please check.

@artlowel 
> There is a point at around 750px wide (you can set that in the device toolbar in the dev tools) where the header breaks:

this was due to the different value of grid breakpoints [we had here](https://github.com/DSpace/dspace-angular/blob/cfbf5871843ae381f17909bfdf39fb997c6800c9/src/styles/_bootstrap_variables.scss#L52-L58). Removing them and using the default values all seems to work properly, a double-check is appreciated.

tdonohue commented:
My apologies @atarix83 , the merging of other high priority PRs caused more merge conflicts here. I believe they all result from merging #1041 (but some may be from #1025).

In the meantime, I'll pull down this PR's branch (prior to merge conflicts) and test that version.

tdonohue reviewed (approved):
👍 @atarix83 : I think this looks great now!  I see the double footer is now gone, the header is one line (and all in the `<header>` element), and I no longer see that "point at around 750px wide..where the header breaks" (previously I was able to reproduce it.

So, I'd recommend we move this forward as-is (_once the merge conflicts are corrected_).  Other minor cleanup / fixes can occur in follow-up PRs.  Thanks again!

artlowel reviewed (approved):
Thanks @atarix83! 

>>There is a point at around 750px wide (you can set that in the device toolbar in the dev tools) where the header breaks:
>
>this was due to the different value of grid breakpoints we had here. Removing them and using the default values all seems to work properly, a double-check is appreciated.

I also don't see any issues when the custom grid breakpoints are removed.

This can be merged when the merge conflicts are solved

atarix83 commented:
@tdonohue merge conflict resolved
